pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -107,6 +108,64 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 		return reconcile.Result{}, err
 	}
 
+	// Finalizer name
+	userAccFinalizerName := "finalizer.toolchain.dev.openshift.com"
+
+	// Add the finalizer if it is not present
+	if !containsString(userAcc.ObjectMeta.Finalizers, userAccFinalizerName) {
+		userAcc.ObjectMeta.Finalizers = append(userAcc.ObjectMeta.Finalizers, userAccFinalizerName)
+		if err := r.client.Update(context.TODO(), userAcc); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+	}
+
+	// Check if the UserAccount has been deleted
+	if !userAcc.ObjectMeta.DeletionTimestamp.IsZero() {
+		reqLogger.Info("deleting UserAccount and subsequent resources", "name", userAcc.Name)
+
+		// Get the Identity associated with the UserAccount
+		identity := &userv1.Identity{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: ToIdentityName(userAcc.Spec.UserID)}, identity); err != nil {
+			if errors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+				// Return and don't requeue
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+
+		// Get the User associated with the UserAccount
+		user := &userv1.User{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name}, user); err != nil {
+			if errors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+				// Return and don't requeue
+				return reconcile.Result{}, nil
+			}
+			return reconcile.Result{}, err
+		}
+
+		// Delete Identity associated with UserAccount
+		if err := r.client.Delete(context.TODO(), identity); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// Delete User associated with UserAccount
+		if err := r.client.Delete(context.TODO(), user); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// Remove finalizer from UserAccount
+		userAcc.ObjectMeta.Finalizers = removeString(userAcc.ObjectMeta.Finalizers, userAccFinalizerName)
+		if err := r.client.Update(context.Background(), userAcc); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		return reconcile.Result{}, nil
```

Comment:
Design change - good to have: Can we move this new_logic out of `Reconcile()` to diff method.  Currently, `Reconcile()` mostly call other method so it can reflect the overall flow rather internal logic.

Reply:
Yes! Makes sense - addressed as part of d8498c421eaf5dd07ecf330e6cdd4063ce57e21d